### PR TITLE
Prefer strongly typed enums

### PR DESCRIPTION
**Pull request by sbrkopac**
## Description

Prefer modern strongly typed enums over the old style of enums. [Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class)

## Type of change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Built against local VSG projects. Upstream projects would need to be updated.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by sbrkopac** [comment]
This is a pretty sweeping change that I will leave as a draft for now. Please note that I haven't made updates to the android platform and there should potentially be some operators defined rather than the ugly casting everywhere.

File: src/vsg/io/VSG.cpp [hunk 1]
```diff
@@ -275,7 +275,7 @@ bool VSG::write(const vsg::Object* object, std::ostream& fout, ref_ptr<const Opt
 
 bool VSG::getFeatures(Features& features) const
 {
-    features.extensionFeatureMap["vsgb"] = static_cast<FeatureMask>(READ_FILENAME | READ_ISTREAM | READ_MEMORY | WRITE_FILENAME | WRITE_OSTREAM);
-    features.extensionFeatureMap["vsgt"] = static_cast<FeatureMask>(READ_FILENAME | READ_ISTREAM | READ_MEMORY | WRITE_FILENAME | WRITE_OSTREAM);
+    features.extensionFeatureMap["vsgb"] = static_cast<FeatureMask>(static_cast<uint8_t>(FeatureMask::READ_FILENAME) | static_cast<uint8_t>(FeatureMask::READ_ISTREAM) | static_cast<uint8_t>(FeatureMask::READ_MEMORY) | static_cast<uint8_t>(FeatureMask::WRITE_FILENAME) | static_cast<uint8_t>(FeatureMask::WRITE_OSTREAM));
```
Comment (robertosfield) [inline src/vsg/io/VSG.cpp:278]: This screams out a regression in readability and usability.  Really WTF.  If you ever wanted an illustration of why some C++ features should be used with care this is it.  Code should be made more expression and readable,

**Comment by robertosfield** [comment]
I understand the allure of this type of change, but doing a blanket implementation is a big fail.

Readability and usability are crucial for C++ applications, if a changes makes something less readable and usable then you have to stop what you are doing and figure out a better way that is more expressive of the intent of the programmer.

**Comment by sbrkopac** [comment]
That's specially why it's a draft pull request and it's noted in the comment that the casting is ugly and should be address. I'll go ahead and close this as there doesn't look like there is much appetite to discuss.

**Comment by robertosfield** [comment]
I'm not closed off on introducing more strongly types enums where appropriate.  In each instance one has to ask does it make the code easier to follow, easier to maintain, more expressive.  

Places where you have to add lots casts or extra code compensate for making an enum more strongly typed is very likely a poor place for strongly type enums.

**Pull request closed without merging**